Convert function test fixtures to JSON

djmattyg007 opened the pull request:
This makes it significantly easier to re-use the fixtures directly in ports to other languages.

djmattyg007 commented:
If you'd like, I can split up the `typeof` checks so that more exact feedback can be put into exception messages.

alexesprit commented:
I think we can get rid of all files from the `test/function` files. Initially I split the [whole](https://github.com/web-scrobbler/web-scrobbler/blob/3ad06b8a6d6219f289dcf3b338c8b5ffbc3c8ffa/tests/content/filter.js) test file to keep fixtures for each function separately. Now it's done by keeping them in JSON files.

So, basically, my idea is following:
```
map = {
    'fixture-name': filterFunctionRef,
}

for (each fixture and function in map) {
     testFilterFunction(function, loadFixtureData(fixture));
}
```

alexesprit commented on test/helper/test-function.ts:
```diff
@@ -41,6 +41,13 @@ export function testFilterFunction(
 	describe(`Test '${functionName}' filter function`, () => {
 		for (const data of testData) {
 			const { description, funcParameter, expectedValue } = data;
+			if (
```
I feel this check should be done in the `loadFixtureFile` function, although it requires 2x loops. Does this change increase executing time significantly?

alexesprit replied:
Well I see `loadFixtureFile` is a generic, so we could make a wrapper which returns the FunctionTestData implementations directly.

djmattyg007 replied:
I think we'd need to be dealing with thousands of test cases before we start noticing a performance impact from processing each test case twice. But yeah, the reason I didn't do it in `loadFixtureFile` was because it was meant to be generic in case other fixture files are added in future. I'd be happy to add a wrapper function that validates the data before passing it to `testFilterFunction`.

And yeah, I was wondering if you'd want to combine everything back into a single file but I forgot to bring it up. I'll make that change too.

djmattyg007 commented:
I just force-pushed a brand new commit that contains the following changes:

- Deleted all existing function test files
- Created new functions.spec.ts file that runs all function test cases for all functions
- Moved function fixtures from `test/fixtures/function` to `test/fixtures/functions`
- Added wrapper function for loading fixtures that performs validation
- Marked function test case interface properties as `readonly` because actual testing code should never need to modify them

alexesprit commented on test/functions.spec.ts:
```diff
@@ -0,0 +1,56 @@
+	'remove-version': removeVersion,
+	'remove-parody': removeParody,
+	'remove-feature': removeFeature,
+	'youtube': youtube,
```
I noticed ESLint shows an error for this line when I run `npm run lint`:
```
51:2  error  Expected property shorthand  object-shorthand
```

 Could you suppress it (for this line or for the entire file - both ways are fine)?

I'm not sure why GA runner does not produce an error. Maybe it's related to https://github.com/web-scrobbler/metadata-filter/commit/e1fa65d53683bd42e78483a67f122d24185a2ad0.

djmattyg007 replied:
I think a better approach might be to change how the rule is configured everywhere. ESLint's `object-shorthand` rule supports a mode named `consistent-as-needed`, which I think would be most appropriate to cover this situation. Thoughts?

alexesprit replied:
Yes, that's better.
